src/structs/common_macros.rs

```diff
@@ -79,156 +129,160 @@ macro_rules! pointwise_sub(
-                $( self.$compN *= right; )+
+            fn $method(self, right: &'a N) -> $t<N> {
+                $t::new(
+                    $(
+                        $imp::$method(&self.$compN, &right)
+                    ),+
+                )
             }
         }
 
-        impl Mul<$t<f32>> for f32 {
-            type Output = $t<f32>;
+        impl<'a, N> $imp<N> for &'a $t<N> where for<'b> &'a N: $imp<&'b N, Output = N> {
+            type Output = $t<N>;
 
             #[inline]
-            fn mul(self, right: $t<f32>) -> $t<f32> {
-                $t::new($(self * right.$compN),+)
+            fn $method(self, right: N) -> $t<N> {
+                $t::new(
+                    $(
+                        $imp::$method(&self.$compN, &right)
+                    ),+
+                )
             }
         }
 
-        impl Mul<$t<f64>> for f64 {
-            type Output = $t<f64>;
+        impl<'a, N> $imp<&'a N> for $t<N> where N: $imp<&'a N, Output = N> {
+            type Output = $t<N>;
 
             #[inline]
-            fn mul(self, right: $t<f64>) -> $t<f64> {
-                $t::new($(self * right.$compN),+)
+            fn $method(self, right: &'a N) -> $t<N> {
+                $t::new(
+                    $(
+                        $imp::$method(self.$compN, &right)
+                    ),+
+                )
             }
         }
-    )
+    }
 );
 
 
-macro_rules! pointwise_scalar_div(
+macro_rules! pointwise_scalar_mul(
     ($t: ident, $($compN: ident),+) => (
-        impl<N: Copy + Div<N, Output = N>> Div<N> for $t<N> {
-            type Output = $t<N>;
-
+        impl<N: Copy + MulAssign<N>> MulAssign<N> for $t<N> {
             #[inline]
-            fn div(self, right: N) -> $t<N> {
-                $t::new($(self.$compN / right),+)
+            fn mul_assign(&mut self, right: N) {
+                $( self.$compN *= right; )+
             }
         }
 
+        ref_binop_scalar!(impl Mul, mul for $t; $($compN),+);
+    )
+);
+
+
+macro_rules! pointwise_scalar_div(
+    ($t: ident, $($compN: ident),+) => (
         impl<N: Copy + DivAssign<N>> DivAssign<N> for $t<N> {
             #[inline]
             fn div_assign(&mut self, right: N) {
                 $( self.$compN /= right; )+
             }
         }
+
+        ref_binop_scalar!(impl Div, div for $t; $($compN),+);
     )
 );
 
 
 macro_rules! pointwise_scalar_add(
     ($t: ident, $($compN: ident),+) => (
-        impl<N: Copy + Add<N, Output = N>> Add<N> for $t<N> {
-            type Output = $t<N>;
-
-            #[inline]
-            fn add(self, right: N) -> $t<N> {
-                $t::new($(self.$compN + right),+)
-            }
-        }
-
         impl<N: Copy + AddAssign<N>> AddAssign<N> for $t<N> {
             #[inline]
             fn add_assign(&mut self, right: N) {
                 $( self.$compN += right; )+
             }
         }
 
-        impl Add<$t<f32>> for f32 {
```

Comment:
This, and all the other implementations of Add/Mul/Div in this file that involve f32 or f64 explicitly must not be removed. They allow the user to write things like: `10.0f32 * v`. The other, generic, implementations allow the scalar to be only on the right-hand-side.

f32 and f64 and mentioned explicitly because it is not possible to write something generic like `impl<N> Mul<Vector3<N>> for N { ... }`